internal/controller/manager/rbac.go

```diff
@@ -0,0 +1,142 @@
+
+import (
+	"text/template"
+
+	"github.com/crunchydata/postgres-operator/internal/config"
+	"github.com/crunchydata/postgres-operator/internal/ns"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// ErrReconcileRBAC defines the error string that is displayed when RBAC reconciliation is
+	// enabled for the current PostgreSQL Operator installation but the Operator is unable to
+	// to properly/fully reconcile its own RBAC in a target namespace.
+	ErrReconcileRBAC = "operator is unable to reconcile RBAC resource"
+)
+
+// ReconcileRBAC is responsible for reconciling  a configMap resource that has obtained from
+// the ConfigMap controller's worker queue
+func (c *ControllerManager) reconcileRBAC(targetNamespace string) {
+
+	log.Debugf("Controller Manager: Now reconciling RBAC in namespace %s", targetNamespace)
+
+	// Use the image pull secrets of the operator service account in the new namespace.
+	operator, err := c.controllers[targetNamespace].kubeClientset.CoreV1().
+		ServiceAccounts(c.pgoNamespace).Get(ns.OPERATOR_SERVICE_ACCOUNT, metav1.GetOptions{})
+	if err != nil {
+		// just log an error and continue so that we can attempt to reconcile other RBAC resources
+		// that are not dependent on the Operator ServiceAccount, e.g. Roles and RoleBindings
+		log.Errorf("%s: %v", ErrReconcileRBAC, err)
+	}
+
+	saCreatedOrUpdated := c.reconcileServiceAccounts(targetNamespace,
+		operator.ImagePullSecrets)
+	c.reconcileRoles(targetNamespace)
+	c.reconcileRoleBindings(targetNamespace)
+
+	// If a SA was created or updated, or if it doesnt exist, ensure the image pull secrets
+	// are up to date
+	for _, reference := range operator.ImagePullSecrets {
+
+		var doesNotExist bool
+
+		if _, err := c.controllers[targetNamespace].kubeClientset.CoreV1().
+			Secrets(targetNamespace).Get(
+			reference.Name, metav1.GetOptions{}); err != nil {
+			if kerrors.IsNotFound(err) {
+				doesNotExist = true
+			} else {
+				log.Errorf("%s: %v", ErrReconcileRBAC, err)
+				continue
+			}
+		}
+
+		if doesNotExist || saCreatedOrUpdated {
+			if err := ns.CopySecret(c.controllers[targetNamespace].kubeClientset, reference.Name,
+				c.pgoNamespace, targetNamespace); err != nil {
+				log.Errorf("%s: %v", ErrReconcileRBAC, err)
+			}
+		}
+	}
+}
+
+// reconcileRoles reconciles the Roles required by the operator in a target namespace
+func (c *ControllerManager) reconcileRoles(targetNamespace string) {
+
+	reconcileRoles := map[string]*template.Template{
+		ns.PGO_TARGET_ROLE:   config.PgoTargetRoleTemplate,
+		ns.PGO_BACKREST_ROLE: config.PgoBackrestRoleTemplate,
+		ns.PGO_PG_ROLE:       config.PgoPgRoleTemplate,
+	}
+
+	for role, template := range reconcileRoles {
+		if err := ns.ReconcileRole(c.controllers[targetNamespace].kubeClientset, role,
+			targetNamespace, template); err != nil {
+			log.Errorf("%s: %v", ErrReconcileRBAC, err)
+		}
+	}
```

Comment:
Since its inception, the namespace package has encapsulated no only namespace functionality, but also the functionality for creating various RBAC resources in the.  So traditionally it has taken that holistic view of a namespace definition, which includes not only the namespace itself, but also the RBAC required by the Operator to operate within that namespace.  So that's essentially what you're still seeing here, i.e. the relationship between namespace and RBAC reconciliation.

I can honestly see it from both perspectives.  I feel as though this patch takes some big strides in further decoupling RBAC stuff from namespace stuff, which implies further separating these things.  But RBAC reconciliation is handled/called by the namespace controller, so there is still a relationship with namespace and RBAC at a technical level there.  

As for whether it should be in the controller, I tend to think of the the controller being the caller some other package containing the business logic for whatever action needs to be performed, and then handling the coordination of all actions that need to be performed based on a specific event.  So in this example the `ns` package knows how to reconcile a RBAC resource, and the controller has the domain specific knowledge to tell it what resources to reconcile and coordinate the reconciliation of all RBAC resources.  So maybe eventually we break it out into an `rbac` package?

Honestly, I think there are various ways to look at this things, and as we continue to refactor this code (which has been a bit of a journey) we can continue to evolve some of these pieces as we see fit.  But since we're already making some big changes here in general, I want to be sure to tackle that incrementally.